File: app/src/main/java/com/example/urlreceiverapp/MainActivity.java
package com.example.urlreceiverapp;

import android.os.Bundle;
import android.webkit.WebView;
import android.webkit.WebViewClient;

import androidx.appcompat.app.AppCompatActivity;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.net.ServerSocket;
import java.net.Socket;
import java.util.concurrent.CopyOnWriteArrayList;

import android.os.Handler;
import android.os.Looper;

public class MainActivity extends AppCompatActivity {


    private WebView webView;
    private final CopyOnWriteArrayList<String> imageUrls = new CopyOnWriteArrayList<>();
    private boolean htmlLoaded = false;
    private final Handler mainHandler = new Handler(Looper.getMainLooper());

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        webView = findViewById(R.id.webView);
        webView.getSettings().setJavaScriptEnabled(true);
        webView.setWebViewClient(new WebViewClient());

        startServer();
    }

    private void startServer() {
        Thread serverThread = new Thread(() -> {
            try (ServerSocket serverSocket = new ServerSocket(11007)) {
                while (true) {
                    Socket socket = serverSocket.accept();
                    BufferedReader reader = new BufferedReader(new InputStreamReader(socket.getInputStream()));
                    String url = reader.readLine();
                    socket.close();

                    if (url != null && !url.trim().isEmpty()) {
                        imageUrls.add(url.trim());
                        updateSlideshow();
                    }
                }
            } catch (Exception e) {
                e.printStackTrace();
            }
        });
        serverThread.start();
    }

    private void updateSlideshow() {
        if (!htmlLoaded) {
            final String htmlContent = "<html>\n" +
                    "<head>\n" +
                    "<style>\n" +
                    "body { margin: 0; background: black; display: flex; justify-content: center; align-items: center; height: 100vh; }\n" +
                    "img { max-width: 100%; max-height: 100%; object-fit: contain; }\n" +
                    "</style>\n" +
                    "</head>\n" +
                    "<body>\n" +
                    "<img id=\"slideshow\" src=\"\" />\n" +
                    "<script>\n" +
                    "let images = [];\n" +
                    "let index = 0;\n" +
                    "function showNextImage() {\n" +
                    "  if (images.length > 0) {\n" +
                    "    document.getElementById(\"slideshow\").src = images[index];\n" +
                    "    index = (index + 1) % images.length;\n" +
                    "  }\n" +
                    "}\n" +
                    "setInterval(showNextImage, 3000);\n" +
                    "</script>\n" +
                    "</body>\n" +
                    "</html>";

            mainHandler.post(() -> {
                webView.loadDataWithBaseURL(null, htmlContent, "text/html", "UTF-8", null);
            });

            htmlLoaded = true;

            new Thread(() -> {
                try {
                    Thread.sleep(1000);
                    injectUrlsToWebView();
                } catch (InterruptedException e) {
                    e.printStackTrace();
                }
            }).start();

        } else {
            injectUrlsToWebView();
        }
    }

    private void injectUrlsToWebView() {
        StringBuilder jsArray = new StringBuilder();
        for (int i = 0; i < imageUrls.size(); i++) {
            jsArray.append("\"").append(imageUrls.get(i)).append("\"");
            if (i < imageUrls.size() - 1) {
                jsArray.append(",");
            }
        }

        final String jsScript = "javascript:images = [" + jsArray.toString() + "];";

        mainHandler.post(() -> {
            webView.evaluateJavascript(jsScript, null);
        });
    }
}

File: app/src/main/java/com/example/urlreceiverapp/MainActivity.kt
package com.example.urlreceiverapp

import android.os.Bundle
import android.webkit.WebView
import android.webkit.WebViewClient
import androidx.appcompat.app.AppCompatActivity
import kotlinx.coroutines.*
import java.io.BufferedReader
import java.io.InputStreamReader
import java.net.ServerSocket
import java.util.concurrent.CopyOnWriteArrayList

class MainActivity : AppCompatActivity() {

    private lateinit var webView: WebView
    private val imageUrls = CopyOnWriteArrayList<String>()
    private var htmlLoaded = false

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        webView = findViewById(R.id.webView)
        webView.settings.javaScriptEnabled = true
        webView.webViewClient = WebViewClient()

        startServer()
    }

    private fun startServer() {
        CoroutineScope(Dispatchers.IO).launch {
            val serverSocket = ServerSocket(11007)

            while (true) {
                val socket = serverSocket.accept()
                val input = BufferedReader(InputStreamReader(socket.getInputStream()))
                val url = input.readLine()
                socket.close()

                if (!url.isNullOrBlank()) {
                    imageUrls.add(url)
                    updateSlideshow()
                }
            }
        }
    }

    private fun updateSlideshow() {
        if (!htmlLoaded) {
            val htmlContent = """
                <html>
                <head>
                    <style>
                        body { margin: 0; background: black; display: flex; justify-content: center; align-items: center; height: 100vh; }
                        img { max-width: 100%; max-height: 100%; object-fit: contain; }
                    </style>
                </head>
                <body>
                    <img id="slideshow" src="" />
                    <script>
                        let images = [];
                        let index = 0;

                        function showNextImage() {
                            if (images.length > 0) {
                                document.getElementById("slideshow").src = images[index];
                                index = (index + 1) % images.length;
                            }
                        }

                        setInterval(showNextImage, 3000);
                    </script>
                </body>
                </html>
            """.trimIndent()

            runOnUiThread {
                webView.loadDataWithBaseURL(null, htmlContent, "text/html", "UTF-8", null)
            }
            htmlLoaded = true

            CoroutineScope(Dispatchers.Main).launch {
                delay(1000)
                injectUrlsToWebView()
            }
        } else {
            injectUrlsToWebView()
        }
    }

    private fun injectUrlsToWebView() {
        val jsArray = imageUrls.joinToString(",") { "\"$it\"" }
        val jsScript = "javascript:images = [$jsArray];"
        runOnUiThread {
            webView.evaluateJavascript(jsScript, null)
        }
    }
}
